pydantic/_internal/_generate_schema.py

```diff
@@ -266,7 +247,11 @@ def _generate_schema_from_property(self, obj: Any, source: Any) -> core_schema.C
             # Can return None to tell pydantic not to override
             return get_schema(source=source, gen_schema=self)
 
-        schema_property = getattr(obj, '__pydantic_core_schema__', None)
+        if not _typing_extra.is_dataclass(obj):
+            schema_property = getattr(obj, '__pydantic_core_schema__', None)
+        else:
+            # Only use the __pydantic_core_schema__ if it is defined on this exact class, not a parent class
+            schema_property = obj.__dict__.get('__pydantic_core_schema__')
```

Comment:
generally better to do

```py
if true_thing:
    true logic
else:
    false logic
```

than

```py
if not true_thing:
    false logic
else:
    true thing
```

(pedantry)

Reply:
I kept it in that order because, for readability, I usually like the first branch to be the more common path, but I'll change it to do the `if true: if false:` order.